### ISSUE TITLE
typeclass and instance heads don't quantify their type variables with forall

### ISSUE THREAD

**Issue by michaelficarra** [purescript/purescript#1120]
I believe these are the only places where we don't require `forall`. Examples from the website:

```purs
class Sized a where
  size :: a -> Number

instance sizedArray :: (Sized a) => Sized [a] where
  size [] = 0
  size (x : xs) = size x + size xs
```

Proposed new requirement:

```purs
class forall a. Sized a where
  size :: a -> Number

instance sizedArray :: forall a. (Sized a) => Sized [a] where
  size [] = 0
  size (x : xs) = size x + size xs
```

**Comment by Fresheyeball** [comment]
@michaelficarra  are you seriously opening issues during PureScript Conf?

**Comment by michaelficarra** [comment]
[Maybe](https://github.com/slamdata/purescript-aff/issues/21). It's inspiring!

**Comment by paf31** [comment]
:+1:

**Comment by garyb** [comment]
Is requiring `forall` a feature? I thought we only did that becasue we hadn't got around to allowing Haskell-style implicit typevars.

**Comment by hdgarrood** [comment]
I find `class forall a. Sized a where [...]` a bit strange - first, you wouldn't be allowed to use a real type in that context (that is, `class Sized Foo where [...]` isn't allowed), and also, I think of `class Sized a where` as introducing a type variable `a` into the inner scope already, so there shouldn't be a need for an extra type-variable-introducing construct (that is, `forall`).

:+1: for the instance part, though, I do find the current behaviour a bit jarring.

**Comment by paf31** [comment]
I agree with @hdgarrood in theory, but in practice, this change is _massive_ for relatively little gain IMO.

**Comment by michaelficarra** [comment]
Also agreed with @hdgarrood. I still think this is an inconsistency worth fixing eventually. It's unfortunate that this got passed over for 0.7.0.

**Comment by jdegoes** [comment]
One could at least allow explicit quantification syntax for a few versions.

**Comment by michaelficarra** [comment]
Ah, good idea. If both are allowed, the current syntax can be deprecated using our new warning system.

**Comment by paf31** [comment]
Backlog?

**Comment by garyb** [comment]
Yeah, I think so.

**Comment by natefaubion** [comment]
I personally would still really like to see this. I find reading implicitly quantified instances difficult, and have even made silly typo mistakes due to this.

**Comment by jrairigh** [comment]
I'd like to work on this feature, but am new to purescript.  @natefaubion does #3824 affect this feature or the other way around?

**Comment by natefaubion** [comment]
@jrairigh It's the other way around. Implementing this feature would likely affect the syntax of #3824.

**Comment by jrairigh** [comment]
I'm guessing the changes would go somewhere in the CST parsing code.  Is this correct?  If not, could someone please point me in the right direction?

**Comment by hdgarrood** [comment]
This would indeed go somewhere in the CST parsing code; probably by adding new cases to `instHead` in `lib/purescript-ast/src/Language/PureScript/CST/Parser.y`. I think we might also want to start emitting warnings for implicitly quantified variables in instances, and then turn that into an error later.

**Comment by jrairigh** [comment]
I've added a new deprecated warning constructor to ParserWarningType called **WarnDeprecatedInstHeadForallSyntax**. What should the deprecation warning text to be?
